src/System.Private.CoreLib/shared/System/Threading/Mutex.Windows.cs

```diff
@@ -163,16 +91,26 @@ private static OpenExistingResult OpenExistingWorker(string name, out Mutex resu
                     return OpenExistingResult.NameNotFound;
                 if (Interop.Errors.ERROR_PATH_NOT_FOUND == errorCode)
                     return OpenExistingResult.PathNotFound;
-                if (null != name && Interop.Errors.ERROR_INVALID_HANDLE == errorCode)
+                if (Interop.Errors.ERROR_INVALID_HANDLE == errorCode)
                     return OpenExistingResult.NameInvalid;
 
                 // this is for passed through Win32Native Errors
                 throw Win32Marshal.GetExceptionForWin32Error(errorCode, name);
             }
 
             result = new Mutex(myHandle);
-
             return OpenExistingResult.Success;
         }
+
+        // Note: To call ReleaseMutex, you must have an ACL granting you
+        // MUTEX_MODIFY_STATE rights (0x0001). The other interesting value
+        // in a Mutex's ACL is MUTEX_ALL_ACCESS (0x1F0001).
+        public void ReleaseMutex()
+        {
+            if (!Interop.Kernel32.ReleaseMutex(_waitHandle))
+            {
+                throw new ApplicationException(SR.Arg_SynchronizationLockException);
```

Comment:
@kouvel I know you have looked into this in the past. Do you think we should use this for the shared copy or should we switch to `ThrowSignalOrUnsignalException` from CoreRT?

Reply:
```ThrowSignalOrUnsignalException``` throws the same error if the _waitHandle is not null
https://github.com/dotnet/corert/blob/86277500c50fd6d83879737fb2b95ad7c98e107a/src/System.Private.CoreLib/src/System/Threading/WaitHandle.Windows.cs#L258

Reply:
ReleaseMutex may result in `ERROR_INVALID_HANDLE` or `ERROR_NOT_OWNER`. Core and Netfx intend to throw ApplicationException(message: SynchronizationLockException) for both cases. However, when the SafeWaitHandle is set to null, since this is accessing the field (which can be null) and not the property (which can't be null), we actually get an ArgumentNullException:

```c#
        var m = new Mutex();
        m.WaitOne();
        m.SafeWaitHandle = null;
        m.ReleaseMutex();
```

There were a whole bunch of differences like these all over the place in the wait handle code when it came to invalid handles. I had tried to turn those cases into a consistent behavior that can be relied upon across APIs (InvalidOperationException), though the exception type is different from Core/Netfx.

Reply:
We do get an ApplicationException when it's actually an invalid handle value:

```c#
        var m = new Mutex();
        m.WaitOne();
        m.SafeWaitHandle = new ManualResetEvent(false).SafeWaitHandle;
        m.ReleaseMutex();
```

It's probably fine to match Core/Netfx wherever reasonable, though there are some weird cases.